design-proposals/klip-26-java-client-interfaces.md

```diff
@@ -0,0 +1,715 @@
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  Integer getInt(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as a long.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  Long getLong(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as a long.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  Long getLong(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as a double.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  Double getDouble(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as a double.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  Double getDouble(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as a boolean.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  Boolean getBoolean(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as a boolean.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  Boolean getBoolean(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as a decimal.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  BigDecimal getDecimal(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as a decimal.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  BigDecimal getDecimal(String columnName);
+}
+```
+
+We considered representing column types in a more structured form (rather than plain strings) to accomodate complex/nested data types but felt this added complexity would not be helpful for most use cases. (This would also require a server-side change, so it's a fair bit of additional work.)
+
+For `getDecimal(...)` in the `Row` interface, rather than trying to parse the column type to extract the precision and scale, we will simply convert the value to a BigDecimal without explicitly specifying the precision and scale. We could also add an option for users to specify the scale and precision in the getter, if we think that would be useful.
+
+We could also add `getList(...)` and `getMap(...)` methods to the `Row` interface, but it's not clear to me how valuable this would be given that the nested data types would not be known.
+
+Tim also proposed switching from generic Java types (`Map<String, Object>`, `List<Object>`) to Vert.x types (`JsonObject`, `JsonArray`) as the latter is more type-safe and the client already has a dependency on Vert.x anyway. The downside, though, is that then apps that use the client would be required to depend on Vert.x as well.
```

Comment:
+1 to exposing `Map` not `Json*` I also prefer exposing interfaces, not classes, so that we have freedom about how we want to implement it underneath the hood. In other systems I've worked on we've done lots of optimizations around the clients in completely backwards compatible ways because we could play around with the return type (e.g. lazily parsing nested fields)

Reply:
Returning rich JsonObject and JsonArray has quite a few advantages - you have type safe getters, methods to serialize to string form and you can get still get it as a map if you want with getMap(). I find it a bit odd that you dislike the untyped string column types, but you prefer an unttyped jsonObject represented as a map ;)

Reply:
While I understand the advantages, I am with @agavra that it seems questionable to expose "third party" classes as public API.

If we (for whatever reason) in the future what to move off Vert.x, all applications using the client would break, if we need to keep some Vert.x dependency to just keep the return types. Overall, this coupling seems no desirable.

Reply:
> I find it a bit odd that you dislike the untyped string column types, but you prefer an unttyped jsonObject represented as a map ;)

My top preference would be what you suggested on the PR (having our types that wrap JsonObject) but that seems a little overkill. I'm not totally sure why JsonObject gives me more type safety than just a Map does - but maybe I'm missing something (if I understand correctly, getMap _does not_ return a JsonObject, it returns a map - the value type is consistent).

Also the concern both here and with `String` above was being able to play around with the implementation as well as type safety

Reply:
>> I'm not totally sure why JsonObject gives me more type safety than just a Map does

Because it has methods getString(), getInt(), getDecimal etc unlike a map where you have check type and cast.

Reply:
>> My top preference would be what you suggested on the PR (having our types that wrap JsonObject) but that seems a little overkill

I'm good with that too, the issue isn't about want to expose vert.x specific classes, it's about providing something that has the most value to the user. Imho providing a rich class which supports nested Json types is a big win, and would be useful for struct values, for example.

Reply:
I would vote to replicate the classes but not expose vert.x specific ones.